### DISCUSSION TITLE
Instanced material equivalent

### DISCUSSION THREAD

**Discussion by jihworks** [stride3d/stride#1626] [Q&A] [answered]
I'm a UE user and currently researching the Stride features for future works.

I read the documents about material and archetype part, but it seems missing description about how they work in C#.

Espeacially, I need to edit the material's attributes such as diffuse color by CPU logics.
In UE, it's done by material [parameters](https://docs.unrealengine.com/5.1/en-US/instanced-materials-in-unreal-engine/).
In low-level graphics, a dynamic constant buffer will work for them.

Can you provide some samples or documents for this?

**Comment by tebjan** [comment, marked as answer]
Stride supports [3 ways of passing instancing transformations](https://github.com/tebjan/StrideTransformationInstancing) to the material.

[The shader language](https://doc.stride3d.net/4.0/en/manual/graphics/effects-and-shaders/shading-language/index.html) supports composition, so each material property can be customized with a shader, for example inheriting from `ComputeColor` or `ComputeFloat`. These shaders can use the `streams.InstanceID` variable to load values from a buffer or contact array. See also the shader sample in the launcher, if you study it, it should be clear how to use it.

To understand the shader inheritance hierarchy, you can use [the shader explorer](https://github.com/tebjan/Stride.ShaderExplorer).

**Reply by jihworks** [reply]
Thanks for reply. But, I decided to use Unity. Unfortunately, I will not proof how it works.
